pkg/operator/feature_controller_set.go

```diff
@@ -0,0 +1,70 @@
+package operator
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift/cert-manager-operator/api/operator/v1alpha1"
+	cm "github.com/openshift/cert-manager-operator/pkg/controller/certmanager"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const retryInterval = 2 * time.Minute
+
+type FeatureControllerSet struct {
+	v1alpha1.FeatureName
+
+	controllers []ManagedController
+	log         logr.Logger
+}
+
+func NewFeatureControllerSet(feature v1alpha1.FeatureName, controllers []ManagedController) *FeatureControllerSet {
+	return &FeatureControllerSet{
+		FeatureName: feature,
+		controllers: controllers,
+	}
+}
+
+func (fc *FeatureControllerSet) run(mgr ctrl.Manager) error {
+	for _, c := range fc.controllers {
+		err := c.SetupWithManager(mgr)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (fc *FeatureControllerSet) pollAndRunIfEnabled(ctx context.Context, mgr ctrl.Manager, feat cm.FeatureAccessor) {
+	shouldEnable := false
+
+	for !shouldEnable {
+		select {
+		case <-ctx.Done():
+			fc.log.V(1).Info("graceful shutdown")
+			return
+
+		case <-time.Tick(retryInterval):
```

Comment:
not a blocker, instead of periodic poll, could we share a channel between `Reconcile` and   `RunWithManagerOnceEnabled` methods, say a new field in `CertManagerReconciler` and as arg to the later method, and wait on the channel for updates?

Reply:
Added an event watcher to the FeatureAccessor, which takes care of doing the similar. Events are sent by the feature accessor when a new feature is enabled, this is subscribed by the `RunWithManagerOnceEnabled` method which triggers enablement accordingly.